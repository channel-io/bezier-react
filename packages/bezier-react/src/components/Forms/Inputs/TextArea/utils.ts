import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
