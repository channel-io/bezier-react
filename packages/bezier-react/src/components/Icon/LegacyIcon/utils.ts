/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
