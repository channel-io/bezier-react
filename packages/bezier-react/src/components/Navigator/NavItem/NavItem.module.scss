.LeftIconWrapper {
  display: flex;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.RightContentWrapper {
  display: flex;
  flex: 1 0 auto;
  align-items: center;
  justify-content: flex-end;
  margin-left: 8px;
}

.Wrapper {
  display: block;
}

.Item {
  display: flex;
  align-items: center;
  height: 28px;
  padding: 0 6px;
  text-decoration: none;
  outline: 0;
  border-radius: var(--radius-6);

  &:where(.active) {
    color: var(--bgtxt-blue-normal);
    background-color: var(--bgtxt-blue-lightest);
  }

  &:where(:not(.active)) {
    color: var(--txt-black-darkest);
    &:hover {
      background-color: var(--bg-black-lighter);
    }

    /*
      ':focus-visible' Pseudo class는 Safari 15.4 이후 지원하기 때문에,
      이를 ':hover' 와 ',' 연산자로 연결할 경우 hover, focus-visible 모두 작동하지 않는 이슈가 있음.
      https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible#browser_compatibility
    */
    &:focus-visible {
      background-color: var(--bg-black-lighter);
    }
  }
}
