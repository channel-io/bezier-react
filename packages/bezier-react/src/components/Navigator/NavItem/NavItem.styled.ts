/* Internal dependencies */
import { styled, css } from 'Foundation'
import type { InterpolationProps } from 'Types/Foundation'

interface WrapperProps extends InterpolationProps {
  active: boolean
}

export const LeftIconWrapper = styled.div`
  display: flex;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
`

export const RightContentWrapper = styled.div`
  display: flex;
  flex: 1 0 auto;
  align-items: center;
  justify-content: flex-end;
  margin-left: 8px;
`

const activeItemStyle = css`
  color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-normal']};
  background-color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-lightest']};
`

const nonActiveItemStyle = css`
  color: ${({ foundation }) => foundation?.theme?.['txt-black-darkest']};
  
  
  &:hover {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }
  
  /*
    ':focus-visible' Pseudo class는 Safari 15.4 이후 지원하기 때문에,
    이를 ':hover' 와 ',' 연산자로 연결할 경우 hover, focus-visible 모두 작동하지 않는 이슈가 있음.
    https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible#browser_compatibility
  */
  &:focus-visible {
    background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
  }
`

export const Item = styled.a<WrapperProps>`
  display: flex;
  align-items: center;
  height: 28px;
  padding: 0 6px;
  text-decoration: none;
  outline: 0;
  ${({ foundation }) => foundation?.rounding?.round6}

  ${({ active }) => (active ? activeItemStyle : nonActiveItemStyle)}

  ${({ interpolation }) => interpolation}
`

export const Wrapper = styled.li`
  display: block;
`
