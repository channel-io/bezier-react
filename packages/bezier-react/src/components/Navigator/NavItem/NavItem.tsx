/* External dependencies */
import React, { useCallback, memo } from 'react'

/* Internal dependencies */
import { Typography } from 'Foundation'
import { noop } from 'Utils/functionUtils'
import { LegacyIcon, IconSize, isIconName } from 'Components/Icon'
import { Text } from 'Components/Text'
import type NavItemProps from './NavItem.types'
import {
  Item,
  Wrapper,
  RightContentWrapper,
  LeftIconWrapper,
} from './NavItem.styled'

export const NAV_ITEM_TEST_ID = 'bezier-react-nav-item'
export const NAV_ITEM_LEFT_ICON_TEST_ID = 'bezier-react-nav-item-left-icon'

function NavItem({
  as,
  testId = NAV_ITEM_TEST_ID,
  name,
  style,
  className,
  interpolation,
  content,
  href,
  target = '_self',
  rightContent,
  leftIcon,
  active,
  onClick = noop,
}: NavItemProps) {
  const handleClickItem = useCallback((e?: React.MouseEvent) => {
    onClick(e, name)
  }, [
    name,
    onClick,
  ])

  const showLeftIcon = isIconName(leftIcon)

  return (
    <Wrapper role="none">
      <Item
        as={as}
        active={active}
        style={style}
        className={className}
        interpolation={interpolation}
        onClick={handleClickItem}
        data-testid={testId}
        href={href}
        target={target}
        role="menuitem"
      >
        <LeftIconWrapper>
          { showLeftIcon && (
            <LegacyIcon
              testId={NAV_ITEM_LEFT_ICON_TEST_ID}
              name={leftIcon}
              size={IconSize.S}
              color={active ? 'bgtxt-blue-normal' : 'txt-black-dark'}
            />
          ) }
        </LeftIconWrapper>

        <Text typo={Typography.Size14} truncated>
          { content }
        </Text>

        { rightContent && (
          <RightContentWrapper>
            { rightContent }
          </RightContentWrapper>
        ) }
      </Item>
    </Wrapper>
  )
}

export default memo(NavItem)
