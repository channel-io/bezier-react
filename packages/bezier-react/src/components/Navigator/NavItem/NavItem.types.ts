import { type BezierIcon } from '@channel.io/bezier-icons'

import type {
  ActivatableProps,
  AlphaBezierComponentProps,
  ContentProps,
  LinkProps,
  SideContentProps,
} from '~/src/types/ComponentProps'

interface NavItemOwnProps {
  name: string
  target?: HTMLAnchorElement['target']
  onClick?: (e?: React.MouseEvent, name?: string) => void
}

export interface NavItemProps extends
  AlphaBezierComponentProps,
  ContentProps,
  LinkProps,
  SideContentProps<BezierIcon, React.ReactNode>,
  ActivatableProps,
  Omit<React.HTMLAttributes<HTMLAnchorElement>, 'onClick' | 'content'>,
  NavItemOwnProps {}
