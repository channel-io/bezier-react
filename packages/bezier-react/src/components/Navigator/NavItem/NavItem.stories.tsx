/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from '~/src/utils/storyUtils'
import { Icon, IconSize, ErrorFilledIcon } from '~/src/components/Icon'
import NavItem from './NavItem'
import type NavItemProps from './NavItem.types'

export default {
  title: getTitle(base),
  component: NavItem,
} as Meta

const Template: Story<NavItemProps> = (args) => (
  <div style={{ width: '240px', padding: '6px' }}>
    <NavItem {...args} />
  </div>
)

export const Primary: Story<NavItemProps> = Template.bind({})

Primary.args = {
  active: false,
  name: 'general',
  content: '일반 설정',
  href: 'https://google.com',
  leftIcon: undefined,
  rightContent: <Icon source={ErrorFilledIcon} size={IconSize.XS} color="bgtxt-orange-normal" />,
}
