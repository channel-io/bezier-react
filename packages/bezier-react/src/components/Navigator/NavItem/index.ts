import NavItem from './NavItem'
import type NavItemProps from './NavItem.types'

export type {
  NavItemProps,
}

export {
  NavItem,
}
