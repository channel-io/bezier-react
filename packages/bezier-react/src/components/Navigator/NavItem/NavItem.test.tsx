/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { LightTheme } from '~/src/foundation/Colors/Theme'
import { render } from '~/src/utils/testUtils'
import NavItem, {
  NAV_ITEM_LEFT_ICON_TEST_ID,
  NAV_ITEM_TEST_ID,
} from './NavItem'
import type NavItemProps from './NavItem.types'

describe('NavItem Test >', () => {
  let props: NavItemProps

  beforeEach(() => {
    props = {
      leftIcon: 'dot',
      name: 'general',
      content: 'test-content',
    }
  })

  const renderNavItem = (optionProps?: Partial<NavItemProps>) =>
    render(<NavItem {...props} {...optionProps} />)

  describe('Snapshot >', () => {
    it('Active', () => {
      const { getByTestId } = renderNavItem({ active: true })

      const rendered = getByTestId(NAV_ITEM_TEST_ID)

      expect(rendered).toMatchSnapshot()
    })

    it('Not active', () => {
      const { getByTestId } = renderNavItem({ active: false })

      const rendered = getByTestId(NAV_ITEM_TEST_ID)

      expect(rendered).toMatchSnapshot()
    })
  })

  describe('LeftIcon Color', () => {
    it('Icon color should be "bgtxt-blue-normal" when active prop is true', () => {
      const { getByTestId } = renderNavItem({ active: true })

      const rendered = getByTestId(NAV_ITEM_LEFT_ICON_TEST_ID)

      expect(rendered).toHaveStyle(`color: ${LightTheme['bgtxt-blue-normal']}`)
    })

    it('Icon color should be "txt-black-dark" when active prop is false', () => {
      const { getByTestId } = renderNavItem({ active: false })

      const rendered = getByTestId(NAV_ITEM_LEFT_ICON_TEST_ID)

      expect(rendered).toHaveStyle(`color: ${LightTheme['txt-black-dark']}`)
    })
  })
})
