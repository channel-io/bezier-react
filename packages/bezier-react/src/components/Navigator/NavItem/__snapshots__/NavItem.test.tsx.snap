// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem Test > Snapshot > Active 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 28px;
  padding: 0 6px;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  overflow: hidden;
  border-radius: 6px;
  color: #5E56F0;
  background-color: #5E56F01A;
}

<a
  class="c0"
  data-testid="bezier-react-nav-item"
  role="menuitem"
  target="_self"
>
  <div
    class="c1"
  >
    <svg
      class="Icon margin"
      data-testid="bezier-react-nav-item-left-icon"
      fill="none"
      height="20"
      style="--b-icon-color: var(--bgtxt-blue-normal);"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="Text typo-14 truncated margin"
    data-testid="bezier-react-text"
  >
    test-content
  </span>
</a>
`;

exports[`NavItem Test > Snapshot > Not active 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 20px;
  max-width: 20px;
  margin-right: 8px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 28px;
  padding: 0 6px;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  overflow: hidden;
  border-radius: 6px;
  color: #000000D9;
}

.c0:hover {
  background-color: #0000000D;
}

.c0:focus-visible {
  background-color: #0000000D;
}

<a
  class="c0"
  data-testid="bezier-react-nav-item"
  role="menuitem"
  target="_self"
>
  <div
    class="c1"
  >
    <svg
      class="Icon margin"
      data-testid="bezier-react-nav-item-left-icon"
      fill="none"
      height="20"
      style="--b-icon-color: var(--txt-black-dark);"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M17 12a5 5 0 1 1-10 0 5 5 0 0 1 10 0"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <span
    class="Text typo-14 truncated margin"
    data-testid="bezier-react-text"
  >
    test-content
  </span>
</a>
`;
